sql/core/src/test/scala/org/apache/spark/sql/Spark18601Suite.scala

```diff
@@ -0,0 +1,331 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.spark.SparkFunSuite
+import org.apache.spark.sql.catalyst.expressions.{Alias, Attribute, Cast, CreateArray, CreateMap, CreateNamedStruct, CreateNamedStructLike, CreateNamedStructUnsafe, Expression, GetArrayItem, GetArrayStructFields, GetMapValue, GetStructField, IntegerLiteral, Literal}
+import org.apache.spark.sql.catalyst.optimizer.Optimizer
+import org.apache.spark.sql.catalyst.plans.PlanTest
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.test.SharedSQLContext
+import org.apache.spark.sql.types._
+import org.scalatest.{FunSuite, ShouldMatchers, Suite, Tag}
+
+import scala.collection.immutable.IndexedSeq
+import org.apache.spark.sql.catalyst.dsl.expressions._
+import org.apache.spark.sql.catalyst.dsl.plans._
+
+/**
+* Created by eyalf on 11/4/2016.
+*/
+class Spark18601Suite extends PlanTest with SharedSQLContext with ShouldMatchers{
+  //import testImplicits._
+
+  lazy val baseRelation = sqlContext.range( 1L, 1000L)
+  lazy val baseOptimizedPlan = baseRelation.queryExecution.optimizedPlan
+
+  val idRef = ('id).long.notNull
+  val idRefColumn = Column( "id" )
+  val struct1RefColumn = Column( "struct1" )
+
+  test( "explicit" ) {
+    val rel = baseRelation.select(
+      functions.struct( idRefColumn as "att" ).getField("att") as "outerAtt"
+    )
+    rel.schema shouldEqual
+      StructType( StructField( "outerAtt", LongType, nullable = false ) :: Nil )
+
+    val optimized = rel.queryExecution.optimizedPlan
+
+    val expected = baseOptimizedPlan.select(
+      idRef as "outerAtt"
+    )
+
+    comparePlans(optimized, expected)
+  }
+
+  test( "explicit - deduced att name") {
+    val rel = baseRelation.select(
+      functions.struct( idRefColumn as "att" ).getField("att")
+    )
+    rel.schema shouldEqual
+      StructType(
+        StructField( "named_struct(att, id AS `att`).att", LongType, nullable = false ) :: Nil
+      )
+
+    val optimized = rel.queryExecution.optimizedPlan
+
+    val expected = baseOptimizedPlan.select(
+      idRef as "named_struct(att, id AS `att`).att"
+    )
+
+    comparePlans(optimized, expected)
+  }
+
+  test( "collapsed" ) {
+    val rel = baseRelation.select(
+      functions.struct( idRefColumn as "att" ) as "struct1"
+    )
+    rel.schema shouldEqual
+      StructType(
+        StructField(
+          "struct1",
+          StructType(
+            StructField(
+              "att",
+              LongType,
+              false
+            ) :: Nil
+          ),
+          false
+        ) :: Nil
+      )

```

Comment:
Kind of hard to read. Could you make it shorter? For example,
```Scala
val mapType = MapType(
  StringType,
  StructType(StructField("att1", LongType, nullable = false) :: Nil),
  valueContainsNull = false)
val expectedType = StructType(StructField("m", mapType, nullable = false) :: Nil)
rel.schema shouldEqual expectedType
```

The same issue for all the other testcases